leveldb/datastore.go

```diff
@@ -0,0 +1,160 @@
+package leveldb
+
+import (
+	"errors"
+	"os"
+	"path/filepath"
+	"sync"
+
+	ds "github.com/ipfs/go-datastore"
+	dsq "github.com/ipfs/go-datastore/query"
+	"github.com/syndtr/goleveldb/leveldb"
+	"github.com/syndtr/goleveldb/leveldb/opt"
+	"github.com/syndtr/goleveldb/leveldb/util"
+	"go.uber.org/atomic"
+)
+
+var (
+	_ ds.Datastore    = (*Datastore)(nil)
+	_ ds.TxnDatastore = (*Datastore)(nil)
+	// ErrClosed is an error message returned when the datastore is no longer open
+	ErrClosed = errors.New("datastore closed")
+)
+
+// Datastore is a go-datastore implement using leveldb
+type Datastore struct {
+	db         *leveldb.DB
+	path       string
+	closed     *atomic.Bool
+	syncWrites bool
+	close      sync.Once
+	closeLock  sync.RWMutex
+}
+
+// Options is an alias of syndtr/goleveldb/opt.Options which might be extended
+// in the future.
+type Options = opt.Options
+
+// NewDatastore returns a new datastore backed by leveldb
+func NewDatastore(path string, opts *Options) (*Datastore, error) {
+	var noSync bool
+	if opts != nil {
+		noSync = opts.NoSync
+	}
+	db, err := leveldb.OpenFile(path, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	ds := Datastore{
+		db:     db,
+		path:   path,
+		closed: atomic.NewBool(false),
+		// if noSync is false this will be true
+		syncWrites: !noSync,
+	}
+	return &ds, nil
+}
+
+// Put stores a key-value pair in leveldb
+func (d *Datastore) Put(key ds.Key, value []byte) (err error) {
+	if d.closed.Load() {
+		return ErrClosed
+	}
+	return d.db.Put(key.Bytes(), value, &opt.WriteOptions{Sync: d.syncWrites})
+}
+
+// Sync is a noop
+func (d *Datastore) Sync(prefix ds.Key) error {
```

Comment:
I think Sync should not be a noop is we have noSync set.

Reply:
So leveldb doesn't really have a sync in the sense of a function you can call. The way sync works is that whenever you write a value to the key-value store, you can specify whether it is a sync or non-sync write